### ISSUE TITLE
Is it possible to have a hybrid shell e.g. Rust+Swift/Kotlin

### ISSUE THREAD

**Issue by ntangldev** [redbadger/crux#165]
I have an iOS project that is already using Rust libraries via FFI.

I have been looking for a solution to better structure this App as it grows. I'm expecting to integrate more Rust libraries.

I took this route to simplify migrating the App to Android and Desktop.

Crux looks like a great solution for the functional core.

The rust libraries I am currently using provide database and file sync capabilities.

I can support the capabilities in Swift and route Requests to the third party Rust libraries where necessary.

This does not seem optimal, is there any support for a hybrid shell with some requests getting routed to rust implementations and some to Swift/Kotlin?

**Comment by StuartHarris** [comment]
Hey there, my sincere apologies for slow response (have been on annual leave). I think the idea of having shell-side capability implementations in Rust is interesting (if I understand you correctly). This would, I guess, raise the abstraction level of the capability, encouraging a fatter shell-side implementation. My personal preference is to make the shell-side implementation as thin (and therefore as generic) as possible. This encourages as much logic as possible to be pushed to the core (or _a_ core, if you want to keep the logic separate/resuable). But I suspect this may not always be practical (if it requires rewriting an existing library, for example, to push the logic inwards and the side-effects outwards). 
I think we need to think a bit about how that might be implemented so that it would still work for Wasm (web) targets.

**Comment by ntangldev** [comment]
Thank you for the response, I was wondering if composable applications might be a useful here, but could not find any examples or docs.

My thought was that nested Apps might support a rust shell or pass the capability request to a parent App.  The outer most App would continue to route unsatisfied capabilities to one of the runtime host shells.

**Comment by StuartHarris** [comment]
Right, got you. Yeah, I think that if the shell-side of a component becomes a mission to write in multiple languages, it may indicate that the capability is too high-level.
I have an [example of an app that nests a reusable core](https://github.com/StuartHarris/crux-passkey), which effectively uses a core as a high-level "pseudo capability" by orchestrating a couple of low-level Capabilities. Does this help?

**Comment by ntangldev** [comment]
Thanks, I'll try that pattern.
